src/main/scala/mesosphere/marathon/core/task/termination/TaskKillService.scala

```diff
@@ -0,0 +1,46 @@
+package mesosphere.marathon.core.task.termination
+
+import mesosphere.marathon.core.task.Task
+
+import scala.concurrent.Future
+
+/**
+  * A service that handles killing tasks. This will take care about extra logic for lost tasks,
+  * apply a retry strategy and throttle kill requests to Mesos.
+  */
+trait TaskKillService {
+  // TODO: should we favor distinct vs overloaded functions? killTask, killTaskById, killTasks, killUnknownTask?
+
+  /**
+    * Kill the given tasks and return a future that is completed when all of the tasks
+    * have been reported as terminal.
+    *
+    * @param tasks the tasks that shall be killed.
+    * @return a future that is completed when all tasks are killed.
+    */
+  def kill(tasks: Iterable[Task]): Future[Unit]
+
+  /**
+    * Kill the given task. The implementation should add the task onto a queue that is processed
+    * short term and will eventually kill the task.
+    *
+    * @param taskId the id of the task that shall be killed.
+    */
+  def kill(taskId: Task.Id): Unit
```

Comment:
tl;dr: the code that kills single tasks in a fire and forget manner doesn't need to wait for results, or does that by subscribing to the eventStream.

The `killTasks(..)` replaces the former `StoppingBehavior` trait that was used by TaskKillActor/AppStopActor which were used when scaling down/stopping an app. In that case, the actors performed the _killing and watching_, and completed a Future when ready. The related implementations around that now use `killTasks` and wait for the Future, leaving waiting, throttling and retrying up to the `TaskKillProgressActor`.

Most other places where Marathon issues kill requests are fire and forget requests. E.g., the `TaskReplaceActor` (which performs app upgrades) kills tasks in that manner and subscribes to the eventBus listening for `MesosStatusUpdateEvent`'s to figure out whether it can kill another task/start a new one/has finished.

So, yes, we can return a Future for all of those, makes sense to me. But then we need to adjust the usages and rely on the futures instead of the event bus which means slightly more changes that I wanted to do here.

We have some work items scheduled around deployment logic and that change would fit just fine I guess.